File: rust/bridge/shared/src/usernames.rs
//
// Copyright 2023 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

#[allow(unused_imports)]
use ::usernames::{
    create_for_username, decrypt_username, NicknameLimits, Username, UsernameError,
    UsernameLinkError,
};
use libsignal_bridge_macros::*;

#[allow(unused_imports)]
use crate::support::*;
use crate::*;

#[bridge_fn]
pub fn Username_Hash(username: String) -> Result<[u8; 32], UsernameError> {
    Username::new(&username).map(|un| un.hash())
}

#[bridge_fn]
pub fn Username_Proof(username: String, randomness: &[u8; 32]) -> Result<Vec<u8>, UsernameError> {
    Username::new(&username)?.proof(randomness)
}

#[bridge_fn]
pub fn Username_Verify(
    proof: &[u8],
    hash: &[u8],
) -> Result<(), ::usernames::ProofVerificationFailure> {
    if hash.len() != 32 {
        return Err(::usernames::ProofVerificationFailure);
    }
    let mut arr = [0u8; 32];
    arr.copy_from_slice(hash);
    Username::verify_proof(proof, arr)
}

#[bridge_fn]
pub fn Username_CandidatesFrom(
    nickname: String,
    min_len: u32,
    max_len: u32,
) -> Result<Box<[String]>, UsernameError> {
    let mut rng = rand::rngs::OsRng;
    let limits = NicknameLimits::new(min_len as usize, max_len as usize);
    Username::candidates_from(&mut rng, &nickname, limits).map(Vec::into_boxed_slice)
}

#[bridge_fn]
pub fn Username_HashFromParts(
    nickname: String,
    discriminator: String,
    min_len: u32,
    max_len: u32,
) -> Result<[u8; 32], UsernameError> {
    let limits = NicknameLimits::new(min_len as usize, max_len as usize);
    Username::from_parts(&nickname, &discriminator, limits).map(|un| un.hash())
}

#[bridge_fn(ffi = false)]
pub fn UsernameLink_Create(
    username: String,
    entropy: Option<&[u8]>,
) -> Result<Vec<u8>, UsernameLinkError> {
    let mut rng = rand::rngs::OsRng;
    let entropy = entropy
        .map(|buf| {
            buf.try_into()
                .map_err(|_| UsernameLinkError::InvalidEntropyDataLength)
        })
        .transpose()?;
    let (entropy, mut buffer) = create_for_username(&mut rng, username, entropy)?;
    buffer.splice(0..0, entropy);
    Ok(buffer)
}

#[bridge_fn(ffi = "username_link_create", jni = false, node = false)]
pub fn UsernameLink_CreateAllowingEmptyEntropy(
    username: String,
    entropy: &[u8],
) -> Result<Vec<u8>, UsernameLinkError> {
    let mut rng = rand::rngs::OsRng;
    let entropy = if entropy.is_empty() {
        None
    } else {
        Some(
            entropy
                .try_into()
                .map_err(|_| UsernameLinkError::InvalidEntropyDataLength)?,
        )
    };
    let (entropy, mut buffer) = create_for_username(&mut rng, username, entropy)?;
    buffer.splice(0..0, entropy);
    Ok(buffer)
}

#[bridge_fn]
pub fn UsernameLink_DecryptUsername(
    entropy: &[u8],
    encrypted_username: &[u8],
) -> Result<String, UsernameLinkError> {
    let entropy = entropy
        .try_into()
        .map_err(|_| UsernameLinkError::InvalidEntropyDataLength)?;
    decrypt_username(entropy, encrypted_username)
}

File: node/ts/usernames.ts
//
// Copyright 2023 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

/* eslint @typescript-eslint/no-shadow: ["error", { "allow": ["hash"] }] */

import { randomBytes } from 'node:crypto';
import { RANDOM_LENGTH } from './zkgroup/internal/Constants.js';
import * as Native from './Native.js';

export type UsernameLink = {
  entropy: Uint8Array;
  encryptedUsername: Uint8Array;
};

export function generateCandidates(
  nickname: string,
  minNicknameLength: number,
  maxNicknameLength: number
): string[] {
  return Native.Username_CandidatesFrom(
    nickname,
    minNicknameLength,
    maxNicknameLength
  );
}

export function fromParts(
  nickname: string,
  discriminator: string,
  minNicknameLength: number,
  maxNicknameLength: number
): { username: string; hash: Uint8Array } {
  const hash = Native.Username_HashFromParts(
    nickname,
    discriminator,
    minNicknameLength,
    maxNicknameLength
  );
  // If we generated the hash correctly, we can format the nickname and discriminator manually.
  const username = `${nickname}.${discriminator}`;
  return { username, hash };
}

export function hash(username: string): Uint8Array {
  return Native.Username_Hash(username);
}

export function generateProof(username: string): Uint8Array {
  const random = randomBytes(RANDOM_LENGTH);
  return generateProofWithRandom(username, random);
}

export function generateProofWithRandom(
  username: string,
  random: Uint8Array
): Uint8Array {
  return Native.Username_Proof(username, random);
}

export function decryptUsernameLink(usernameLink: UsernameLink): string {
  return Native.UsernameLink_DecryptUsername(
    usernameLink.entropy,
    usernameLink.encryptedUsername
  );
}

export function createUsernameLink(
  username: string,
  previousEntropy?: Uint8Array
): UsernameLink {
  const usernameLinkData = Native.UsernameLink_Create(
    username,
    previousEntropy ?? null
  );
  const entropy = usernameLinkData.subarray(0, 32);
  const encryptedUsername = usernameLinkData.subarray(32);
  return { entropy, encryptedUsername };
}

// Only for testing. Will throw on failure.
export function verifyProof(proof: Uint8Array, hash: Uint8Array): void {
  Native.Username_Verify(proof, hash);
}
